### PR TITLE
Apply monitoring linters

### PR DESCRIPTION
**Pull request by assafad**
### What this PR does
Apply the following linters:
- [monitoringlinter](https://github.com/kubevirt/monitoring/tree/main/monitoringlinter) that was implemented in https://github.com/kubevirt/monitoring/pull/221, and was designed to enforce https://github.com/kubevirt/community/pull/219.
This linter ensures that all monitoring-related practices are implemented within the `pkg/monitoring` directory using [operator-observability](https://github.com/machadovilaca/operator-observability/tree/main) methods. It verifies that all metrics, alerts and recording rules registrations are centralized in this directory, and restricts the direct use of [Prometheus registration methods](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Registerer).
Example for a message reported by the linter:

  ```
  kubevirt/pkg/virt-controller/watch/workload-updater/workload-updater.go:67:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
  ```
- [operator-observability rules linter](https://github.com/machadovilaca/operator-observability/tree/main/pkg/testutil) - ensures that alerts and recording rules definitions are following the monitoring best practices, e.g. alerts have all the required annotations and labels with expected values, recording rules have expressions, etc. In addition, bump operator-observability for it.

Fixes https://issues.redhat.com/browse/CNV-36761

### Special notes for your reviewer:
Depending on #11184 and https://issues.redhat.com/browse/CNV-39013.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/monitoring/rules/rules_test.go
Comment (machadovilaca) [inline pkg/monitoring/rules/rules_test.go:22]:
```suggestion
		linter.AddCustomAlertValidations(
			testutil.ValidateAlertNameLength,
			testutil.ValidateAlertRunbookURLAnnotation,
			testutil.ValidateAlertHealthImpactLabel,
			testutil.ValidateAlertPartOfAndComponentLabels
		)

```

File: pkg/monitoring/rules/rules_test.go
Comment (machadovilaca) [inline pkg/monitoring/rules/rules_test.go:27]: I think this is possible:

```suggestion
		problems := linter.LintAlerts(alerts)
		Expect(problems).To(BeEmpty())
```

File: pkg/monitoring/rules/rules_test.go
Comment (machadovilaca) [inline pkg/monitoring/rules/rules_test.go:35]: same as above: `LintRecordingRules`

File: pkg/monitoring/rules/rules_test.go
Comment (machadovilaca) [inline pkg/monitoring/rules/rules_test.go:6]: nit:

```suggestion
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	
	"github.com/machadovilaca/operator-observability/pkg/testutil"
```

**Review by machadovilaca — CHANGES_REQUESTED** [review]

File: pkg/monitoring/rules/rules_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_test.go]: Done

File: pkg/monitoring/rules/rules_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_test.go]: I somehow missed it, thanks, updated.

File: pkg/monitoring/rules/rules_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_test.go]: Done, thanks

File: pkg/monitoring/rules/rules_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_test.go]: Done

File: pkg/monitoring/rules/rules_test.go
Comment (EdDev) [inline pkg/monitoring/rules/rules_test.go:15]: `Expect(SetupRules("")).To(Succeed())`

File: pkg/monitoring/rules/rules_test.go
Comment (EdDev) [inline pkg/monitoring/rules/rules_test.go:1]: Can you add the license header?

File: pkg/monitoring/rules/rules_suite_test.go
Comment (EdDev) [inline pkg/monitoring/rules/rules_suite_test.go:1]: Can you add the license header?

File: pkg/monitoring/rules/rules_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_test.go]: Thanks, done.

File: pkg/monitoring/rules/rules_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_test.go:17]: Done

File: pkg/monitoring/rules/rules_suite_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_suite_test.go:17]: Done

File: pkg/monitoring/rules/rules_suite_test.go
Comment (EdDev) [inline pkg/monitoring/rules/rules_suite_test.go:2]: 2024

File: pkg/monitoring/rules/rules_test.go
Comment (EdDev) [inline pkg/monitoring/rules/rules_test.go:2]: 2024

**Review by EdDev — APPROVED** [review]

File: pkg/monitoring/rules/rules_suite_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_suite_test.go]: Done

File: pkg/monitoring/rules/rules_test.go
Comment (assafad) [reply pkg/monitoring/rules/rules_test.go]: Done

**Review by EdDev — APPROVED** [review]

File: Makefile
Comment (EdDev) [inline Makefile:232]: You should probably add this to the builder image.

Comment (EdDev) [reply Makefile:232]: I mean here: https://github.com/kubevirt/kubevirt/blob/9d36c93dc57f165807e1d339d7dffd8f5fd82231/hack/builder/Dockerfile

**Comment by assafad** [comment]
Since the monitoring refactoring is still WIP (should be completed by https://github.com/kubevirt/kubevirt/pull/11184), currently the monitoring linter gives:
```
monitoringlinter ./pkg/...
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/client/prometheus/prometheus.go:95:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/client/prometheus/prometheus.go:96:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/client/prometheus/prometheus.go:97:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/domainstats/prometheus/prometheus.go:584:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:79:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:80:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:81:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:82:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:83:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:84:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:85:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/reflector/prometheus/prometheus.go:86:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/workqueue/prometheus/prometheus.go:54:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/workqueue/prometheus/prometheus.go:66:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/workqueue/prometheus/prometheus.go:79:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/workqueue/prometheus/prometheus.go:92:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/workqueue/prometheus/prometheus.go:104:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/workqueue/prometheus/prometheus.go:117:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/monitoring/workqueue/prometheus/prometheus.go:132:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
/home/aadmi/git-repos/upstream/kubevirt/pkg/virt-controller/watch/workload-updater/workload-updater.go:67:2: monitoring-linter: metrics should be registered only within pkg/monitoring directory, using operator-observability packages.
make: *** [Makefile:233: lint-monitoring] Error 3
```

File: Makefile
Comment (assafad) [reply Makefile:232]: Don't think this is possible until we complete the refactoring (also adding `make lint-monitoring` to the CI), right?

Comment (EdDev) [reply Makefile:232]: I do not think they are related.
You can push it into the builder in a separate PR, I am not sure if the maintainers will allow you to add this `go install` here.

Comment (assafad) [reply Makefile]: Done, and now we can see the linter errors on https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11454/pull-kubevirt-code-lint/1765752677913858048.
@EdDev can you PTAL again and let me know if anything else is missing from your POV?

Comment (EdDev) [reply Makefile]: Very nice, thanks.

So you plan now to add the fixes to this PR as well? (to pass the linter failures)

**Comment by iholder101** [comment]
General direction seems great to me.
Thanks for your efforts @assafad! Great work!

/approve

However, since I'm not familiar with all of the details, I'd like others that were involved to have a look and have time to give feedback. Let's hold the PR until the involved people review.

/hold

@sradco @machadovilaca @nunnatsa @RamLavi - would you be able to review?
FYI @enp0s3 @vladikr

File: Makefile
Comment (assafad) [reply Makefile]: Thanks, it will be completed by https://github.com/kubevirt/kubevirt/pull/11184 and https://issues.redhat.com/browse/CNV-39013. This PR will be on hold until then.

**Comment by victortoso** [comment]
I'd suggest adding in the `Special notes for your reviewer:` of the first commit that this PR depends on https://github.com/kubevirt/community/pull/219 (and others?) and moving it to Draft while those dependencies don't get merged.

**Comment by assafad** [comment]
/unhold
metrics refactoring is completed. We can now merge this.

**Pull request merged**
